src/ModuleLoader.ts

```diff
@@ -217,7 +220,16 @@ export class ModuleLoader {
 		timeStart('load modules', 3);
 		let source: string | SourceDescription;
 		try {
-			source = (await this.pluginDriver.hookFirst('load', [id])) ?? (await readFile(id));
+			source =
+				(await this.pluginDriver.hookFirst('load', [id])) ??
+				(await this.readQueue.run(async () => {
+					this.graph.parallelFileReads++;
```

Comment:
Rather than storing this as graph specific state, it should be possible to maintain it as module level state since the process is a global concept anyway.

Reply:
The first thing I tried was to globally patch fs with graceful-fs but that lead to some weird behavior and still didn't work. But one possibility is certainly to just replace fs by graceful-fs in `src/utils/fs.ts`. I just tried it and it seems to work fine.
The drawback of that approach would be that the end user has no control over it, there is no way to exactly configure the old behavior, but that might not be an issue, because, as I said, too much concurrency when reading files probably has not advantage anyway.

So in the end it's a decision between
1. Custom solution with configurability
+ flexible
- more code to test and maintain
2. graceful-fs
+ safe, on line diff.
- One more dependency

I don't really have a preference one way or the other as long as we can get rid of the error. 😁 

If you want to have the custom solution, I can adjust according to your comment and keep the state in a different place.
Opinions?

Reply:
While a global concept would be closer to what is intended here, I believe the current implementation is much easier to handle as different builds are isolated from each other, and the option is passed per build. This avoids unintended side effects. Otherwise if I run a build with a concurrency of 1 together with a build of unlimited concurrency, either both would have 1 or both would be unlimited. I think the current solution is fine in that regard.
Regarding graceful-fs, this one seems to do much more than we need here, so I fear it also blows up the build unnecessary and also may have unintended interactions with plugins etc. But decisions can of course always be revisited.

Reply:
Removed the state from graph. I followed @lukastaegert 's sugestions for testing and it's not needed anymore.